File: leetcode.com/balanced-binary-tree.py
#!/bin/python3
# https://leetcode.com/problems/balanced-binary-tree/

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def get_height(self, node: Optional[TreeNode]) -> int:
        if not node:
            return -1

        return 1 + max(self.get_height(node.left), self.get_height(node.right))
        
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True

        return abs(self.get_height(root.left) - self.get_height(root.right)) <= 1 \
            and self.isBalanced(root.left) and self.isBalanced(root.right)

File: leetcode.com/balanced-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Boolean}
def is_balanced(root)
    def depth(node)
        return 0 if node.nil?
        return [depth(node.left), depth(node.right)].max + 1
    end

    def balanced?(node)
        return true if node.nil?

        lh = depth(node.left)
        rh = depth(node.right)

        return false if (lh - rh).abs > 1
        return balanced?(node.left) && balanced?(node.right)
    end

    return balanced?(root)
end